pyClient/zeth/joinsplit.py

```diff
@@ -0,0 +1,680 @@
+from __future__ import annotations
+import zeth.constants as constants
+import zeth.errors as errors
+from zeth.utils import get_trusted_setup_dir, hex_extend_32bytes, \
+    hex_digest_to_binary_string, string_list_flatten, encode_single, \
+    encode_abi, encrypt, decrypt, get_public_key_from_bytes
+from zeth.prover_client import ProverClient
+from api.util_pb2 import ZethNote, JoinsplitInput, HexPointBaseGroup1Affine, \
+    HexPointBaseGroup2Affine
+from nacl.public import PrivateKey, PublicKey  # type: ignore
+import nacl.utils  # type: ignore
+import api.prover_pb2 as prover_pb2
+
+import os
+import json
+import sys
+from Crypto import Random
+from hashlib import blake2s, sha256
+from py_ecc import bn128 as ec
+from typing import Tuple, Dict, List, Iterable, Union, Any
+
+
+FQ = ec.FQ
+G1 = Tuple[ec.FQ, ec.FQ]
+
+
+class ApkAskPair:
+    def __init__(self, a_sk: str, a_pk: str):
+        self.a_pk = a_pk
+        self.a_sk = a_sk
+
+
+SK = Tuple[int, int]
+VK = Tuple[G1, G1]
+
+
+class VkSkPair:
+    def __init__(self, x: int, y: int, x_g1: G1, y_g1: G1):
+        self.vk = (x_g1, y_g1)
+        self.sk = (x, y)
+
+
+# Dictionary representing a VerificationKey from any supported snark
+GenericVerificationKey = Dict[str, Any]
+
+
+# Dictionary representing a Proof from any supported snark
+GenericProof = Dict[str, Any]
+
+
+class NoteRandomness:
+    def __init__(self, rho: str, trap_r: str):
+        self.rho = rho
+        self.trap_r = trap_r
+
+    @staticmethod
+    def new() -> NoteRandomness:
+        """
+        Compute the note randomness: the trapdoor trapR and rho. Starting the
+        Non-Malleability update, rho is computed from phi (see above), the rho
+        generated in this function is thus obsolete except for dummy input
+        notes.
+        """
+        rho = bytes(Random.get_random_bytes(32)).hex()
+        trap_r = bytes(Random.get_random_bytes(48)).hex()
+        return NoteRandomness(rho, trap_r)
+
+
+def create_zeth_note(
+        randomness: NoteRandomness,
+        recipient_apk: str,
+        value: str) -> ZethNote:
+    """
+    We follow the formatting of the proto file. Create a ZethNote description
+    Starting the Non-Malleability update, this function is used only for dummy
+    input notes as rhoS are now structured ( rho = PRF_{phi}(i, phi, h_sig) ).
+    """
+    note = ZethNote(
+        apk=recipient_apk,
+        value=value,
+        rho=randomness.rho,
+        trap_r=randomness.trap_r
+    )
+    return note
+
+
+def create_zeth_notes(
+        phi: str,
+        hsig: str,
+        recipient_apk0: str,
+        value0: str,
+        recipient_apk1: str,
+        value1: str) -> Tuple[ZethNote, ZethNote]:
+    """
+    Create two ordered ZethNotes. This function is used to generate new output
+    notes.
+    """
+    rho0 = compute_rho_i(phi, hsig, 0)
+    randomness0 = NoteRandomness.new()
+    note0 = ZethNote(
```

Comment:
I think some refactoring will be needed here. We can modify the `create_zeth_note` function and invoke it in this one

Reply:
Ididn't see this until I'd already refactored as above.  Let me know if that works.

Reply:
We can keep it that way (ie: remove the `create_zeth_note` as you did). What I was suggesting was to create each notes in this function by calling the `create_zeth_note`, but what you did works, let's keep it that way.